syntax = "proto3";

enum PushStatus {
    SUCCESS = 0;
    FAILURE = 1;
}

message Message {
    string key = 1;
    bytes value = 2;
}

message MessageList {
    int32 replica_id = 1;
    repeated Message messages = 2;
}

message ReplicaRequest {
    int32 id = 1;
    string url = 2;
}

message ReplicaID {
    int32 id = 1;
}

message MessageCount {
    int32 replica_id = 1;
    int32 count = 2;
}

message PushResponse {
    PushStatus status = 1;
}

message Empty {}

service Broker {
    rpc Ack(Empty) returns (Empty) {}

    rpc Push(Message) returns (PushResponse) {}
    rpc Pull(Empty) returns (Message) {}

    rpc SetReplica(ReplicaRequest) returns (Empty) {}
    rpc LeadReplica(ReplicaID) returns (Empty) {}

    rpc PushReplica(MessageList) returns (PushResponse) {}
    rpc DropReplicaMessages(MessageCount) returns (Empty) {}
}